stackslib/src/chainstate/nakamoto/coordinator/mod.rs

```diff
@@ -0,0 +1,720 @@
+/// already been processed.
+pub fn get_nakamoto_reward_cycle_info<U: RewardSetProvider>(
+    burn_height: u64,
+    sortition_tip: &SortitionId,
+    burnchain: &Burnchain,
+    chain_state: &mut StacksChainState,
+    sort_db: &SortitionDB,
+    provider: &U,
+) -> Result<Option<RewardCycleInfo>, Error> {
+    let epoch_at_height = SortitionDB::get_stacks_epoch(sort_db.conn(), burn_height)?
+        .expect(&format!(
+            "FATAL: no epoch defined for burn height {}",
+            burn_height
+        ))
+        .epoch_id;
+
+    assert!(
+        epoch_at_height >= StacksEpochId::Epoch30,
+        "FATAL: called a nakamoto function outside of epoch 3"
+    );
+
+    if !burnchain.is_in_prepare_phase(burn_height) {
+        return Err(Error::NotInPreparePhase);
+    }
+
+    // calculating the reward set for the _next_ reward cycle
+    let reward_cycle = burnchain
+        .block_height_to_reward_cycle(burn_height)
+        .expect("FATAL: no reward cycle for burn height")
+        + 1;
+
+    // only proceed if we have not yet calculated the PoX reward info for this reward cycle.
+    let last_processed_reward_cycle = {
+        let ic = sort_db.index_handle(sortition_tip);
+        ic.get_last_processed_reward_cycle()?
+    };
+
+    if last_processed_reward_cycle >= reward_cycle {
+        return Err(Error::RewardSetAlreadyProcessed);
+    }
+
+    debug!("Processing reward set for Nakamoto reward cycle";
+          "burn_height" => burn_height,
+          "reward_cycle" => reward_cycle,
+          "reward_cycle_length" => burnchain.pox_constants.reward_cycle_length,
+          "prepare_phase_length" => burnchain.pox_constants.prepare_length);
+
+    // find the last Stacks block processed in the preceeding prepare phase
+    // (i.e. the parent of the first Stacks block processed in the prepare phase).
+    // Note that we may not have processed it yet.  But, if we do find it, then it's
+    // unique (and since Nakamoto Stacks blocks are processed in order, the anchor block
+    // cannot change later).
+    let prepare_phase_sortitions =
+        find_prepare_phase_sortitions(sort_db, burnchain, sortition_tip)?;
+    for sn in prepare_phase_sortitions.into_iter() {
+        if !sn.sortition {
+            continue;
+        }
+
+        // find the first Stacks block processed in the prepare phase
+        let Some(prepare_start_block_header) =
+            NakamotoChainState::get_nakamoto_tenure_start_block_header(
+                chain_state.db(),
+                &sn.consensus_hash,
+            )?
+        else {
+            // no header for this snapshot (possibly invalid)
+            continue;
+        };
+
+        let parent_block_id = &prepare_start_block_header
+            .anchored_header
+            .as_stacks_nakamoto()
+            .expect("FATAL: queried non-Nakamoto tenure start header")
+            .parent_block_id;
```

Comment:
This imposes a requirement on the Nakamoto start -- the prepare phase before the first reward cycle must _also_ be evaluated with nakamoto rules. I think this is important to highlight in the comments.

Reply:
Ah, that's an oversight.  Elsewhere, the code assumes that the Nakamoto epoch kicks off early in the reward phase of a reward cycle (e.g. 100 blocks in), so the last Stacks 2.x prepare phase can choose a PoX anchor and give the signers some time to do DKG and prepare to sign Nakamoto blocks.

Reply:
Okay, as of right now (e6df731) the Nakamoto epoch would start in the reward phase of an epoch 2.x reward cycle.